staging/src/k8s.io/apiserver/pkg/storage/testing/watcher_tests.go

```diff
@@ -844,6 +845,177 @@ func TestNamespaceScopedWatch(ctx context.Context, t *testing.T, store storage.I
+		expected            bool
+		allowWatchBookmarks bool
+	}{
+		{ // test old client won't get Bookmark event
+			name:                "allowWatchBookmarks=false",
+			timeout:             3 * time.Second,
+			expected:            false,
+			allowWatchBookmarks: false,
+		},
+		{
+			name:                "allowWatchBookmarks=true",
+			timeout:             3 * time.Second,
+			expected:            true,
+			allowWatchBookmarks: true,
+		},
+	}
+
+	for i, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			pred := storage.Everything
+			pred.AllowWatchBookmarks = tt.allowWatchBookmarks
+			ctx, cancel := context.WithTimeout(context.Background(), tt.timeout)
+			defer cancel()
+
+			watcher, err := store.Watch(ctx, key, storage.ListOptions{ResourceVersion: startRV, Predicate: pred})
+			if err != nil {
+				t.Fatalf("Unexpected error: %v", err)
+			}
+			defer watcher.Stop()
+
+			// Create events of pods in a different namespace
+			out := &example.Pod{}
+			obj := &example.Pod{ObjectMeta: metav1.ObjectMeta{Name: "foo", Namespace: fmt.Sprintf("other-ns-%d", i)}}
+			objKey := computePodKey(obj)
+
+			if err := store.Create(ctx, objKey, obj, out, 0); err != nil {
+				t.Fatalf("Create failed: %v", err)
+			}
+
+			// Now wait for Bookmark event
+			select {
+			case event, ok := <-watcher.ResultChan():
+				if !ok && tt.expected {
+					t.Errorf("Unexpected object watched (no objects)")
+				}
+				if tt.expected && event.Type != watch.Bookmark {
+					t.Errorf("Unexpected object watched %#v", event)
+				}
+			case <-time.After(time.Second * 3):
+				if tt.expected {
+					t.Errorf("Unexpected object watched (timeout)")
+				}
+			}
+		})
+	}
+}
+
+// RunOptionalTestWatchBookmarksWithCorrectResourceVersion tests whether bookmark events are
+// sent with correct resource versions.
+// This feature is currently implemented in watch cache layer, so this is optional.
+//
+// TODO(#109831): ProgressNotify feature is effectively implementing the same
+//
+//	functionality, so we should refactor this functionality to share the same input.
+func RunTestOptionalWatchBookmarksWithCorrectResourceVersion(ctx context.Context, t *testing.T, store storage.Interface) {
+	// Compute the initial resource version.
+	list := &example.PodList{}
+	storageOpts := storage.ListOptions{
+		Predicate: storage.Everything,
+		Recursive: true,
+	}
+	if err := store.GetList(ctx, "/pods", storageOpts, list); err != nil {
+		t.Errorf("Unexpected error: %v", err)
+	}
+	startRV := list.ResourceVersion
+
+	key := "/pods/test-ns"
+	pred := storage.Everything
+	pred.AllowWatchBookmarks = true
+
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
+
+	watcher, err := store.Watch(ctx, key, storage.ListOptions{ResourceVersion: startRV, Predicate: pred, Recursive: true})
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
+	defer watcher.Stop()
+
+	done := make(chan struct{})
+	errc := make(chan error, 1)
```

Comment:
as this is buffered with a small capacity, when we call t.Fatal on L989 do you know what happens to goroutines stuck sending on L975? will they be able to call `wg.Done()` if they can't send their error to the channel?

Reply:
Note that there is exactly one goroutine. And if it puts something into `errc` in L975, immediately after it returns.
So there isn't a possibility for deadlock here.